GPII-2338: Windows service.

stegru opened the pull request:
Here's the Windows Service, that starts GPII and the listeners when a user logs into windows, and restarts them if they die.

At the moment, this provides no value; GPII will operate the same as before. ~~The installer needs to be adjusted to install and start the service.~~

More info in the [README](https://github.com/stegru/windows/blob/GPII-2338/service/README.md).

**Update:**
This changes the way GPII is started. Instead of using the AutoRun registry key, the service starts GPII when the user logs on. Its passed an environment value `GPII_SERVICE_PIPE` which instructs gpii to connect to the pipe (GPII-2399).

If it's ok if GPII is started normally - it just wont be able to enjoy the goodness of the service (currently nothing but restart on crash). However, if the service is started from within the source code (ie, via node service/index.js), then it will open an insecure pipe to which GPII will connect.

The service writes a log to `C:\ProgramData\GPII\gpii-service.log` (`%APPDATA%/GPII/` is not used, because that's the user's profile directory, and the service doesn't run as the user)

stegru commented:
ok to test

stegru commented:
closing for now - don't need CI to check it

stegru commented:
Needs work

This assumes the listeners exist, but when https://github.com/GPII/windows/pull/157 is in they wont.

javihernandez commented:
@stegru let us know when this is ready for review ;)

stegru commented:
aye aye, captain!

stegru commented:
It's ready now... I just have to make a little change, because it currently starts the listeners - and they're about to be chopped.

stegru commented:
The "little change" is done, and this thing is ready for a thorough review!

stegru commented:
ok to test

amb26 commented on gpii/node_modules/serviceHandler/package.json:
```diff
@@ -0,0 +1,13 @@
+{
+    "name": "serviceHandler",
```
Rename this and the directory to gpii-service-handler , it'll avoid massive diffs later

stegru replied:
done

amb26 commented on gpii/node_modules/serviceHandler/src/actions.js:
```diff
@@ -0,0 +1,86 @@
+ * Sends a request to the service.
+ *
+ * @param {Component} service The gpii.serviceHandler instance.
+ * @param {string} action The request action.
```
Capital S, Capital O - proper linting is coming soon for this project from @the-t-in-rtf

stegru replied:
done - and fixed all other offenses

amb26 commented on gpii/node_modules/serviceHandler/src/actions.js:
```diff
@@ -0,0 +1,86 @@
+ *
+ * @param {Component} that The gpii.windows.service.actions instance.
+ * @param {string} command The command to run.
+ * @param {string[]} args Arguments to pass.
```
Looks like we've decided to use Array<String> for these

amb26 replied:
What I wrote was Array left angle bracket String right angle bracket

stegru replied:
done

but what will happen if node introduces generic types...

amb26 commented on gpii/node_modules/serviceHandler/src/requestHandler.js:
```diff
@@ -0,0 +1,66 @@
+    gradeNames: ["fluid.component" ],
+    invokers: {
+        handleRequest: {
+            funcName: "gpii.windows.service.echo",
```
These should probably just be fluid.notImplemented - do we really want a default impl that does nothing?

stegru replied:
That's an actual function, but the name was wrong (but "worked" because the effect of the real function was to call that function).

amb26 commented on gpii/node_modules/serviceHandler/src/serviceHandler.js:
```diff
@@ -0,0 +1,279 @@
+    if (that.reconnect) {
+        // Throttle the retry attempts by a few seconds for each failure, unless the last connection was above 30 seconds.
+        var duration = that.connectTime && process.hrtime(that.connectTime);
+        if (duration && duration[0] > 30) {
```
Stick all these magic numbers into options

amb26 commented on gpii/node_modules/serviceHandler/src/requestHandler.js:
```diff
@@ -0,0 +1,66 @@
+var gpii = fluid.registerNamespace("gpii");
+fluid.registerNamespace("gpii.windows.service.serviceHandler");
+
+// Handles requests from the service.
```
Add some comments here to explain the expected usage pattern - that it is meant to have a bunch of invokers which each handle actions of that name, etc. Seems a bit odd though, in that these seem to have a pretty peculiar contract. Probably implementing these as events plus a bit of bracketing boilerplate to implement the contract would be clearer.

stegru replied:
I hear what you're saying. However, I don't believe events are suitable because there can be zero to many handlers, and the caller expects a result.

As for the boilerplate, there's (now) currently no invoker that requires any parameters. But yes, future ones will pass individual arguments, rather than an object.

amb26 commented on gpii/node_modules/serviceHandler/src/requestHandler.js:
```diff
@@ -0,0 +1,66 @@
+ * Handle a request from the service, by calling the relevant invoker of gpii.windows.service.requestHandler.
+ *
+ * @param {Component} that The gpii.serviceHandler instance.
+ * @param {Object} request The request. It should at least contain "name" which is the name of the request.
```
Looks like you have decided to call this "action" instead. Could you document this structure with a JSDocs @typedef or so

stegru replied:
now `requestName`, added a typedef.

amb26 commented on service/tests/all-tests.js:
```diff
@@ -0,0 +1,48 @@
+ */
+"use strict";
+
+// The service tests are ran in a separate process to the rest of GPII. This not only ensures it's isolated from GPII,
```
Is this going to be a problem when we try to collect code coverage from it?

amb26 commented on service/src/gpiiClient.js:
```diff
@@ -0,0 +1,168 @@
+                        stderr: ""
+                    };
+                    child.stdout.on("data", function (data) {
+                        // Limit the output to ~1 million characters
```
64K?

stegru replied:
The comment is correct - it's the code that was wrong.

amb26 commented on service/src/gpiiClient.js:
```diff
@@ -0,0 +1,168 @@
+ * @param {IpcConnection} ipcConnection The IPC connection.
+ */
+gpiiClient.connected = function (ipcConnection) {
+    this.ipcConnection = ipcConnection;
```
Explain somewhere that this function randomly scrawls on an eventEmitter in this way. Best to document what "this" is using @this as well, and that you modify it.

stegru replied:
an error - fixed.

amb26 commented on service/src/service.js:
```diff
@@ -0,0 +1,159 @@
+ *
+ * @param {String} name Module name
+ * @param {Object} initial [optional] An existing object to add on to.
+ * @return {Object}
```
What object ?

stegru replied:
commented

amb26 commented on service/src/service.js:
```diff
@@ -0,0 +1,159 @@
+ *
+ * See also: https://msdn.microsoft.com/library/ms683241
+ *
+ * @param controlName Name of the control code.
```
Types?

stegru replied:
commented

amb26 commented on service/src/service.js:
```diff
@@ -0,0 +1,159 @@
+    windows = require("./windows.js"),
+    parseArgs = require("minimist");
+
+// Different parts of the service are isolated, and will communicate by emitting events through this central "service"
```
Explain what it actually does, and how it is used - e.g. the service.module entry point, and how each one is given an event, how it is an EventEmitter in itself, and what these events mean, and how these relate to the "event" on a module, etc.

amb26 commented on service/src/winapi.js:
```diff
@@ -0,0 +1,457 @@
+    Struct = require("ref-struct"),
+    arrayType = require("ref-array");
+
+//var ArrayType = arrayType;
```
Axe?

stegru replied:
gone

amb26 commented on service/src/windows.js:
```diff
@@ -0,0 +1,442 @@
+                        break;
+                    case winapi.constants.WAIT_FAILED:
+                        // GetLastError will not work, because WaitForMultipleObjects is called in a different thread.
+                        // Call WaitForMultipleObjects again, but in this thread (with a short timeout in case it works)
```
Yes, it would be pretty awkward if it worked

stegru replied:
What's the worst that could happen? it works

amb26 commented on service/src/gpii-ipc.js:
```diff
@@ -0,0 +1,438 @@
+/**
+ * A connection to a client.
+ * @typedef {Object} IpcConnection
+ * @property authenticate true if authentication is required.
```
Types

stegru replied:
done.

amb26 commented on service/src/gpii-ipc.js:
```diff
@@ -0,0 +1,438 @@
+
+        // Duplicate the event handle for the child.
+        var eventHandleBuf = ref.alloc(winapi.types.HANDLE);
+        var ownProcess = -1 >>> 0;
```
Just write it out in hex rather than alarming the reader

stegru replied:
even added a comment to soften the blow.

amb26 commented on gpii/node_modules/serviceHandler/src/actions.js:
```diff
@@ -0,0 +1,86 @@
+var gpii = fluid.registerNamespace("gpii");
+fluid.registerNamespace("gpii.windows.service.serviceHandler");
+
+// Handles requests to the service.
```
All of the components/members in this hierarchy are pretty generically named and it is pretty hard to work out their relationship by following the threads - "action" "service" "request" "execute" "sendRequest" etc. A little sketch would help, especially explaining why the implementations are held in one component while they are executed on another. Some more specific names would help. Somehow "sendRequest" and "execute" are implemented at the same level, whereas one forwards to the other, and whereas the former looks like it will pass your introspection test on requestHandler.js line 49 and so suggests that it is intended to be used directly.

stegru replied:
One component sends requests, the other receives them. I've commented/renamed/documented to improve understanding.

stegru commented on gpii/node_modules/serviceHandler/src/requestHandler.js:
```diff
@@ -0,0 +1,66 @@
+ */
+gpii.windows.service.handleRequest = function (that, request) {
+    fluid.log("Service: request", request);
+    if (that.requestHandler.options.invokers.hasOwnProperty(request.action)) {
```
I was trying to defend against something like "toString" being passed, but now that I've stepped back... the service just wont do that.

(the "else" is "do nothing").

stegru commented:
ok @amb26, I think I've addressed all your points.

In addition
- There's now an `npm start` which also starts the service. There's also `npm run` helpers in `service/package.json`
- Documented the: IPC [IPC.md](https://github.com/stegru/windows/blob/GPII-2338/service/doc/IPC.md)
- This branch points to a test branch of universal, which has a hacked on-screen keyboard solution which executes a command (`whoami /all`) via the service (and gpii process).
    - `npm start` as a normal non-admin user.
    - Login as `mary`
    - `c:\tmp\whoami-admin.txt` should contain the privileges for the service.
    - `c:\tmp\whoami-self.txt` should contain the privileges of the user. 

(this branch doesn't depend on any univeral changes)

amb26 commented on service/src/gpiiClient.js:
```diff
@@ -0,0 +1,235 @@
+
+"use strict";
+
+var Promise = require("bluebird"),
```
Do we need it? Presumably we can just use ES6 Promises

stegru replied:
No need - done.

amb26 commented on gpii/node_modules/gpii-service-handler/README.md:
```diff
@@ -0,0 +1,19 @@
+# gpiii-service-handler
+
+Module to handing the communications between this GPII user process and the GPII Windows service.
```
Module to handing? We want you the formula TO LONDON BRING!

stegru replied:
fixed

amb26 commented on gpii/node_modules/gpii-service-handler/src/requestSender.js:
```diff
@@ -0,0 +1,118 @@
+fluid.defaults("gpii.windows.service.requestSender", {
+    gradeNames: ["fluid.component" ],
+    invokers: {
+        // called from within.
```
A bit more detail here?

stegru replied:
added

amb26 commented on gpii/node_modules/gpii-service-handler/src/serviceHandler.js:
```diff
@@ -0,0 +1,289 @@
+/*
+ * Windows Service interface.
```
Bit more description will be helpful here what this does!

stegru replied:
yes

amb26 commented on service/package.json:
```diff
@@ -0,0 +1,35 @@
+{
+    "name": "morphic-service",
```
This should still be gpii-service, to match our usage elsewhere, and the directory name should match too

amb26 commented on service/README.md:
```diff
@@ -0,0 +1,170 @@
+## How it works
+
+Services are slightly different to normal processes, the differences are handled by
+[stegru/node-os-service#GPII-2338](https://github.com/stegru/node-os-service/tree/GPII-2338), which is a fork of
```
Please publish a scoped npm dev release of this since git dependencies can cause an arse for npm in the long run. Please also add a touch of explanation here why the fork was needed.

amb26 commented on service/README.md:
````diff
@@ -0,0 +1,170 @@
+When installing the service, add the debug arguments using the `--nodeArgs`. For example:
+
+```
+node index.js --install --nodeArgs=--inspect=0.0.0.0:1234,--debug-brk
````
Presumably this might say --inspect-brk, since debug-brk was deprecated in Node 7

stegru replied:
yes

amb26 commented on service/tests/windows-tests.js:
```diff
@@ -0,0 +1,553 @@
+                e instanceof Error || e.isError);
+            runTest(testIndex + 1);
+        });
+
```
Some funny stretches of whitespace in this file, especially at the bottom

stegru replied:
gone

amb26 commented on gpii/node_modules/gpii-service-handler/src/requestHandler.js:
```diff
@@ -0,0 +1,88 @@
+ * @return {Object|Promise} The return of the handler function.
+ */
+gpii.windows.service.handleRequest = function (that, request) {
+    fluid.log("Service: request", request);
```
I get zero coverage for this function

amb26 commented:
From the child process window I get the following warning:
````
(node:9220) [DEP0005] DeprecationWarning: Buffer() is deprecated due to security and usability issues. Please use the Buffer.alloc(), Buffer.allocUnsafe(), or Buffer.from() methods instead.
````

amb26 commented:
Doesn't there need to be a gpii-app variant of this pull which includes the work to actually restart the Morphic UI app when it dies rather than just the local GPII windows process?

amb26 commented on gpii/node_modules/gpii-service-handler/src/serviceHandler.js:
```diff
@@ -0,0 +1,291 @@
+    var pipeProblem = function (err) {
+        fluid.log("Service: Unable to authenticate: ", err || "Pipe closed");
+        if (promise.disposition) {
+            fluid.log("oops");
```
This could be improved a touch

amb26 commented on service/config/service.json5:
```diff
@@ -0,0 +1,17 @@
+{
+    "processes": {
+        "gpii": {
+            "command": "morphic-app.exe",
```
This one is a bit odd - shouldn't this be configured as part of the gpii-app's module, given it refers to its executable name and also its config file?

javihernandez commented on service/config/service.json5:
```diff
@@ -0,0 +1,17 @@
+            "autoRestart": true,
+            "disabled": false,
+            env: {
+                NODE_ENV: "app.production"
```
No need for setting the NODE_ENV variable, however, we need to specify the GPII_CLOUD_URL as follows:

```diff
diff --git a/service/config/service.json5 b/service/config/service.json5
index 4ae9b57..67b4f33 100644
--- a/service/config/service.json5
+++ b/service/config/service.json5
@@ -7,7 +7,7 @@
             "autoRestart": true,
             "disabled": false,
             env: {
-                NODE_ENV: "app.production"
+                GPII_CLOUD_URL: "https://flowmanager.prd.gcp.gpii.net"
             }
         }
     },
```

javihernandez commented:
ok to test

stegru commented:
ok to test

```
11:09:39.261:  jq: FAIL: Module "Windows Display Settings Handler Tests" Test name "Testing setScreenDpi " - Message: setScreenDpi 0
```

amb26 commented on gpii/node_modules/userListeners/src/windowsLogin.js:
```diff
@@ -0,0 +1,122 @@
+ * by just knowing the sid, without also knowing the signing key.
+ *
+ * @param {Component} that The gpii.windows.userListeners.proximity instance.
+ * @param {Function} sign The function used to sign the user id.
```
Explain signature of function - perhaps with @typedef if necessary

stegru replied:
yes

amb26 commented on gpii/node_modules/userListeners/src/windowsLogin.js:
```diff
@@ -0,0 +1,122 @@
+ * @return {String} The SID of the currently logged in user.
+ */
+gpii.windows.getUserSid = function () {
+    // Lazily initialise the .NET function.
```
Why do we initialise it lazily?

stegru replied:
Added comment `to save having to compile the assembly on start-up as it might not be used.`

amb26 commented on gpii/node_modules/userListeners/src/windowsLogin.js:
```diff
@@ -0,0 +1,122 @@
+/**
+ * Generates a gpiiKey based on the current Windows user.
+ *
+ * The key is generated using the signed digest of the user's SID. This will ensure that a user's key can't be guessed
```
I think we need something simpler than this - an impl that just returns the secret. We can't really afford to have a key which is unstable wrt. the secret, since if this is compromised/revoked, every key changes.

stegru replied:
Right! I just wanted to prevent ids from being easy to generate - but of course once it's compromised it can't be "revoked", and would be next to useless. Instead, I'll just use the per-site domain.

By the way, this isn't "the secret" used for auth, that's in this commit: https://github.com/GPII/windows/pull/155/commits/989f992ed107b6b58297e670be4563365aa01d53 - the interesting bit is in gpii/node_modules/gpii-service-handler/src/serviceHandler.js

I used the simple "just give me the credentials" method as suggested.

stegru commented on gpii/node_modules/gpii-service-handler/src/serviceHandler.js:
```diff
@@ -0,0 +1,332 @@
+        onServiceReady: "{serviceHandler}.events.onConnected"
+    },
+    distributeOptions: {
+        // Make the accessRequester use the service to get the client credentials.
```
@cindyli - could you just double check this is ok (it seems to work). It targets https://github.com/GPII/universal/blob/master/gpii/node_modules/flowManager/src/SettingsDataSource.js#L125 to use the grade at the end of this file (line 303).

So, instead the accessRequester using pilot.json, it asks the service for the `{ client_id, client_secret }`.

cindyli replied:
This distribution looks great. Thanks, @stegru.

amb26 commented on gpii-service/package.json:
```diff
@@ -0,0 +1,34 @@
+        "ffi-napi": "2.4.4",
+        "json5": "2.1.0",
+        "minimist": "1.2.0",
+        "os-service": "stegru/node-os-service#GPII-2338",
```
Let's move this to a scoped npm package published under the GPII org - git dependencies are likely to lay up obscure trouble for us. I've forked the project for us here - https://github.com/GPII/node-os-service
Also, once we aren't rushing for a release, we should submit whatever the required fixes are as a pull to the upstream project. For now, could you list what these fixes are in the README of your forked branch.

stegru replied:
https://github.com/GPII/node-os-service/pull/1

amb26 commented:
Currently got conflicts with master

javihernandez commented on gpii-service/README.md:
```diff
@@ -0,0 +1,170 @@
+
+        /* More processes */
+        "rfid-listener": {
+            "command": "../listeners/GPII_RFIDListener.exe",
```
I guess these don't apply anymore

amb26 replied:
Bump comment - can this be axed?

amb26 commented on gpii/node_modules/userListeners/src/windowsLogin.js:
```diff
@@ -0,0 +1,131 @@
+
+var gpii = fluid.registerNamespace("gpii");
+
+// The proximity user listener
```
It's not the proximity user listener!

amb26 commented:
Hi there @stegru - do you think you could make the update to this branch to support the work in https://github.com/GPII/universal/pull/795 described in the review comments there? Cheers!

stegru commented:
Your wish is my command. https://github.com/GPII/universal/pull/798

stegru commented:
ok to test

Failures unrelated to this.

(looks like a re-occurrence of https://issues.gpii.net/browse/GPII-3543)

```
02:53:09.372:  jq: FAIL: Module "Windows Display Settings Handler Tests" Test name "Testing setScreenDpi " - Message: get/setScreenDpi should return the same
02:53:09.372:  jq: Expected: {
    "configured": 0,
    "minimum": 0,
    "maximum": 1,
    "actual": 0
}
02:53:09.373:  jq: Actual: {
    "configured": 0,
    "minimum": 0,
    "maximum": 0,
    "actual": 0
}
```

stegru commented:
ok to test

Unrelated, potentially random, failure.

```
09:23:37.565:  jq: FAIL: Module "gpii.tests.acceptance.windows.builtIn.config tests" Test name "Testing os_common_highContrast - magnifier running on startup" - Message: Confirming initial state - at sequence position 9 of 30
09:23:37.566:  jq: Expected: {
    "gpii.windows.enableRegisteredAT": {
```

amb26 commented on gpii-service/package.json:
```diff
@@ -0,0 +1,34 @@
+        "ffi-napi": "2.4.4",
+        "json5": "2.1.0",
+        "minimist": "1.2.0",
+        "@gpii/os-service": "stegru/node-os-service#GPII-2338",
```
Still needs to be referenced as a published npm package

stegru replied:
https://www.npmjs.com/package/@gpii/os-service

amb26 commented on package.json:
```diff
@@ -17,7 +19,7 @@
         "edge-js": "10.3.1",
         "ffi-napi": "2.4.3",
-        "gpii-universal": "JavierJF/universal#GPII-3810",
+        "gpii-universal": "stegru/universal#GPII-2338",
```
This pull was merged, you should be able to revert to current master

stegru replied:
done

amb26 commented on gpii-service/shared/pipe-messaging.js:
```diff
@@ -0,0 +1,285 @@
+        if (err instanceof Error) {
+            // Error doesn't serialise
+            e = {};
+            Object.getOwnPropertyNames(err).forEach(function (a) {
```
Object.assign?

stegru replied:
It doesn't work (but you get to keep the test case I just created for it).

```
> Object.assign({}, new Error("antranig"))
{}
```

stegru commented:
ok to test

This is due to a call to SPI, probably taking long (>10 seconds) to broadcast a message to every window, and receive the response. 

```
08:17:23.064:  SPI call has failed
08:17:23.065:  An error occurred in an item of the lifecyclemanager's action queue(invoker: start, func: {that}.processStart), so clearing action queue. Error was {
    "isError": true,
    "message": "Timed out waiting for the SPI call to complete (see GPII-2001)"
}
```

stegru commented:
ok to test

```
09:10:26.566:  jq: FAIL: Module "Windows Display Settings Handler Tests" Test name "Testing setScreenDpi " - Message: setScreenDpi 1
09:10:26.566:  jq: Expected: {
    "actual": 1,
    "configured": 1,
    "minimum": 0,
    "maximum": 1
}
09:10:26.566:  jq: Actual: {
    "configured": 1,
    "minimum": 0,
    "maximum": 0,
    "actual": 0
}
```

amb26 commented:
Test failure in the actual service tests - looks like it might be something real?
````
Service tests ended: 3221225477
10:26:35.106:  jq: FAIL: Module "GPII service tests" Test name "Test window service" - Message: Service tests should pass
10:26:35.106:  jq: Expected: 0
10:26:35.106:  jq: Actual: 3221225477
10:26:35.106:  jq: Source:     at Object.assertEquals (\\vboxsvr\vagrant\node_modules\infusion\tests\test-core\jqUnit\js\jqUnit.js:137:19)
    at ChildProcess.<anonymous> (\\vboxsvr\vagrant\gpii-service\tests\all-tests.js:47:16)
    at ChildProcess.emit (events.js:182:13)
    at maybeClose (internal/child_process.js:962:16)
    at Process.ChildProcess._handle.onexit (internal/child_process.js:251:5)
10:26:35.137:  jq: Test concluded - Module "GPII service tests" Test name "Test window service": 0/1 passed - FAIL
10:26:35.155:  jq: ***************
10:26:35.155:  jq: All tests concluded: 3091/3092 total passed in 155685ms - FAIL
````

amb26 commented:
ok to test

amb26 commented:
Consistent failure:
````
Service tests ended: 3221225477
06:39:02.780:  jq: FAIL: Module "GPII service tests" Test name "Test window service" - Message: Service tests should pass
06:39:02.780:  jq: Expected: 0
06:39:02.998:  jq: Actual: 3221225477
06:39:02.999:  jq: Source:     at Object.assertEquals (\\vboxsvr\vagrant\node_modules\infusion\tests\test-core\jqUnit\js\jqUnit.js:137:19)
    at ChildProcess.<anonymous> (\\vboxsvr\vagrant\gpii-service\tests\all-tests.js:47:16)
    at ChildProcess.emit (events.js:182:13)
    at maybeClose (internal/child_process.js:962:16)
    at Socket.stream.socket.on (internal/child_process.js:381:11)
    at Socket.emit (events.js:182:13)
    at Pipe._handle.close (net.js:606:12)
06:39:03.026:  jq: Test concluded - Module "GPII service tests" Test name "Test window service": 0/1 passed - FAIL
06:39:03.047:  jq: ***************
06:39:03.047:  jq: All tests concluded: 3091/3092 total passed in 202964ms - FAIL
06:39:03.047:  jq: ***************
npm ERR! code ELIFECYCLE
````
@stegru, could you investigate? Also, please try to bind against current universal release - 0.3.0-dev.20191211T142911Z.0af24fabe

stegru commented:
`3221225477` is `0xC0000005`, `STATUS_ACCESS_VIOLATION` (memory access)

ok to test - see how repeatable it is

stegru commented:
ok to test
